### PR TITLE
PCRE2 chartables leakfix

### PR DESCRIPTION
**Pull request by carenas**
This series is a candidate reroll for cb/pcre2-chartables-leakfix, that
hopefully addresses the root cause of the problem reported by Dscho in
Windows, where the PCRE2 library wasn't aware of the custom allocator and
was returning a pointer created with the system malloc but passing it to
NED's free, resulting in a segfault.

The most likely reason why it was triggered by the original leak fix is the layering
violation reported by René and that is likely exclusive to PCRE2 (hence why it
hasn't been reported with PCRE1).  Additional work might be available
in a future release of PCRE2 to address that as detailed in the upstream
bug[1] report.

Changes since v4
* git log change reverted, still not sure where it will fit better and worst case will leak a few bytes when -P is used.  since the users of this API are doing it indirectly it might be problematic long term though, but luckily since it is most of the tine a NOOP and can be called multiple times might be ok to do it unconditionally
* slightly better looking code

Changes since v3 (mostly in patch 2):

* git log also calls the "destructor" for grep API
* no more "bug" being triggered by `make test`, sorry René
* hopefully no more crashes in windows (I was expecting at most a BUG)

Future work (other than the needed refactoring explained in the
second patch) and adjacent bugs, includes:

* tracking more possible users of the grep API that might need to call
  grep_destroy()
* completely moving PCRE2 to use NED (as is done with PCRE1 and was
  proposed on the original patch[2] this is based on
* build on top of the new API so that other work could be shared
  (for example the chartables that started this whole mess)

or (hopefully not)

* ignore the original leak (maybe with an UNLEAK) as René suggested [3]
* discard this work and just use Dscho's fix (with some improvements, like using xmalloc)

[1] https://bugs.exim.org/show_bug.cgi?id=2429
[2] https://public-inbox.org/git/3397e6797f872aedd18c6d795f4976e1c579514b.1565005867.git.gitgitgadget@gmail.com/
[3] https://public-inbox.org/git/7ec60d57-9940-35f2-f7b5-c87d4dc7cdd5@web.de/

### CODE REVIEW

File: grep.c [hunk 1]
```diff
@@ -16,6 +16,22 @@ static int grep_source_is_binary(struct grep_source *gs,
 
 static struct grep_opt grep_defaults;
 
+#ifdef USE_LIBPCRE2
+static pcre2_general_context *pcre2_global_context;
+
+#ifdef USE_NED_ALLOCATOR
```
Comment (dscho) [inline grep.c]: Please do drop any mention of `nedmalloc` in this PR. There is nothing specific to `nedmalloc` about overriding PCRE2's allocator with whatever Git uses, whether the allocator was overridden at build time or at runtime does not matter one bit.

**Comment by dscho** [comment]
@carenas is this PR still live, or do you develop things outside of this branch?

**Comment by carenas** [comment]
> @carenas is this PR still live, or do you develop things outside of this branch?

this was using the obsoleted approach (matching what went to pu) which has been also abandoned upstream AFAIK

I might have the new approach that included Rene's changes somewhere in another pull request but wasn't ready for merging yet and will need to be updated before i can be reviewed as well (ex: rene wanted me to keep his contribution intact and not squash on it a small needed fix) which I might not published yet anywhere

**Comment by dscho** [comment]
@carenas okay. That means this PR can be closed, right?

**Pull request closed without merging**